enhancements/compliance-operator.md

```diff
@@ -0,0 +1,222 @@
+* Provide a way to collect the detailed results of the scan
+  * The results would list a way to remediate the gaps, typically by the means
+    of deploying certain manifests to OKD (e.g. a Machineconfig), a different
+    way of changing cluster configuration or even a free-form recommendation text.
+
+### Non-Goals
+
+The operator itself does not perform any scans or remediations, only
+wraps the OpenSCAP scanner and provides an OKD-native API to
+run the scans.
+
+## Proposal
+
+An instance of a scan would be represented by a CRD. The CRD would include
+the scan profile (e.g. FedRAMP moderate), the content to scan against
+(typically `ocp4`) and optionally a set of individual rules (e.g. is `auditd`
+enabled on the nodes?). The CRD would be processed by the operator, which
+would launch a pod on each node to scan that node and deliver scan results.
+
+The pod would consist of two containers - a scanner and a log collector.
+This is because OpenSCAP writes the report to a file, so the scanner and
+the collector share a directory where the scanner writes the report, the
+collector would wait for the report to be created and write the contents
+of that file to a location where the cluster administrator can grab the
+report and inspect it.
+
+The scanner container in the pod would mount the node's file system into
+a directory and point the `openscap-chroot` binary there.
+
+### User Stories
+
+#### As an OKD admin, I want to assess the degree of compliance of my cluster with FedRAMP moderate.
+
+This is a use-case where the admin just wants to visualize the gaps in
+compliance. They would perform the remediation themselves, potentially
+based on the MCOs the compliance operator suggests.
+
+### Implementation Details
+
+There [exists a PoC](https://github.com/jhrozek/compliance-operator) of the implementation.
+
+Some things to note from the PoC implementation are:
+ * The operator itself schedules a pod per node in the cluster. This was selected
+   (instead of e.g. DaemonSet) because the pods don't have to be long-lived and
+   can just exit when the scan is done.
+ * The collector container in the pod reads the contents of the scan results
+   from a file produced by the scanner container and uploads them to a ConfigMap.
+   This is perhaps a bit hacky, but most volume types do not support
+   ReadWriteMany access mode and we need to get the results from files in
+   N pods across N nodes. We could mount a per-pod volume and then gather
+   the results into another volume when the scan is done, but this still
+   means we need a post-run gather operation.
+ * For viewing the results by administrator might use [a script](https://github.com/jhrozek/scapresults-k8s/blob/master/scapresults/fetchresults.py)
+
+### Risks and Mitigations
+ * [jhrozek]: Not sure what to write here. The scan might be
+   resource-intensive perhaps? The scanner containers need to mount the
+   host fs? The permissions need careful review?
```

Comment:
There are security risks that we need to start considering:

* the operator will need to mount the whole filesystem from the host, we need to make sure that mounting it as read-only is fine to lock down the operator's workload.
* If remediations need to be done, something needs to have permissions to create the MachineConfigs or whatever resources need to be created... we need to come up with a safe scheme to do this, so we don't have an all-powerful operator that could be compromised. Any suggestions on this?

Reply:
@runcom Hi Antonio, do you have any thoughts on the second point?

Reply:
@runcom still would welcome your thoughts on the remediations, if you have any :)

Reply:
> If remediations need to be done, something needs to have permissions to create the MachineConfigs or whatever resources need to be created... we need to come up with a safe scheme to do this, so we don't have an all-powerful operator that could be compromised. Any suggestions on this?

I think that if remediation has to be done automatically then access to the MC's would be needed (service account possibly). If the remediation could be done with the compliance operators help then the compliance operator could generate the MC and pass it to the admin in one way or another. Think selinux and audit log.